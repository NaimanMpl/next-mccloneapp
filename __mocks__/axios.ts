import mockAxios from 'jest-mock-axios';

export const SESSION_MOCK = {
  user: {
    id: "1",
    admin: false,
    email: "johndoe@domain.com",
    name: 'John Doe',
    profileIconUrl: 'awesomeprofilepicture.com/pp/1234',
    role: {
    id: 1,
    name: 'Joueur',
    score: 0
    },
    skin: 'awesomeskin.com/skin/1234'
  }
}

export default mockAxios;