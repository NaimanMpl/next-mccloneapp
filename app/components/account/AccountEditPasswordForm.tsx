'use client';
import { useEditPasswordForm } from '@/app/hooks/useEditPasswordForm';
import { Button } from '@/components/ui/button';
import { Card, CardContent, CardDescription, CardFooter, CardHeader, CardTitle } from '@/components/ui/card';
import { Form, FormControl, FormField, FormItem, FormLabel, FormMessage } from '@/components/ui/form';
import { Input } from '@/components/ui/input';
import { Label } from '@/components/ui/label';
import { Loader2 } from 'lucide-react';
import { useSession } from 'next-auth/react';
import React from 'react';
import AccountPageSkeleton from './AccountPageSkeleton';

const AccountEditPasswordForm = () => {

  const { data: session } = useSession();
  const { form, onSubmit, loading } = useEditPasswordForm();

  return (
    <>
      {!session && <AccountPageSkeleton />}
      {
        session && 
        <Card>
          <CardHeader>
            <CardTitle>Mot de passe</CardTitle>
            <CardDescription>Lorem ipsum dolor sit amet, consectetur adipisicing elit. Molestiae, natus!</CardDescription>
          </CardHeader>
          <Form {...form}>
            <form onSubmit={form.handleSubmit(onSubmit)}>
              <CardContent className='flex flex-col gap-4'>
                <FormField 
                  control={form.control}
                  name='oldPassword'
                  render={({ field }) => (
                    <FormItem>
                      <FormLabel>Ancien mot de passe</FormLabel>
                      <FormControl>
                        <Input className='mt-3 w-80' type='password' placeholder='•••••••••••••' {...field} />
                      </FormControl>
                      <FormMessage />
                    </FormItem>
                  )}
                />
                <FormField 
                  control={form.control}
                  name='newPassword'
                  render={({ field }) => (
                    <FormItem>
                      <FormLabel>Nouveau mot de passe</FormLabel>
                      <FormControl>
                        <Input className='mt-3 w-80' type='password' placeholder='•••••••••••••' {...field} />
                      </FormControl>
                      <FormMessage />
                    </FormItem>
                  )}
                />
                <FormField 
                  control={form.control}
                  name='confirmNewPassword'
                  render={({ field }) => (
                    <FormItem>
                      <FormLabel>Confirmation du mot de passe</FormLabel>
                      <FormControl>
                        <Input className='mt-3 w-80' type='password' placeholder='•••••••••••••' {...field} />
                      </FormControl>
                      <FormMessage />
                    </FormItem>
                  )}
                />
              </CardContent>
              <CardFooter className='border-t border-border px-6 py-2'>
                <div className='flex justify-between items-center w-full'>
                  <CardDescription>Le mot de passe doit au moins contenir 8 caractères.</CardDescription>
                  <Button type='submit' disabled={loading}>
                    {loading && <Loader2 className='mr-2 h-4 w-4 animate-spin' />}
                    Sauvegarder
                  </Button>
                </div>
              </CardFooter>
            </form>
          </Form>
        </Card>
      }
    </>
  )
}

export default AccountEditPasswordForm