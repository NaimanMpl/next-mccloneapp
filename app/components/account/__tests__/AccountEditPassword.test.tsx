import { SESSION_MOCK } from "@/__mocks__/axios";
import { render, screen } from "@testing-library/react";
import user from '@testing-library/user-event';
import { useSession } from "next-auth/react";
import AccountEditPasswordForm from "../AccountEditPasswordForm";

jest.mock('next-auth/react');

describe('<AccountEditPassword />', () => {

  beforeEach(() => {
    (useSession as jest.Mock).mockReturnValue({ data: SESSION_MOCK, update: jest.fn() });
  });

  it('should render successfully', () => {

    const form = render(
      <AccountEditPasswordForm />
    );

    expect(form).toMatchSnapshot();

  });

  it('should not submit the form if password doesnt match', async () => {

    const oldPassword = 'secret';
    const newPassword = 'secretpassword';
    const newConfirmPassword = 'notamatchingpassword';

    render(
      <AccountEditPasswordForm />
    );

    const [ oldPasswordInput, newPasswordInput, confirmNewPasswordInput ] = screen.getAllByPlaceholderText('•••••••••••••');
    const submitButton = screen.getByRole('button');

    await user.type(oldPasswordInput, oldPassword);
    await user.type(newPasswordInput, newPassword);
    await user.type(confirmNewPasswordInput, newConfirmPassword);

    await user.click(submitButton);

  });

});