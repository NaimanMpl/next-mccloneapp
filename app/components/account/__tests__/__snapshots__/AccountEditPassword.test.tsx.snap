// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccountEditPassword /> should render successfully 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="rounded-lg border border-border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="flex flex-col space-y-1.5 p-6"
        >
          <h3
            class="text-2xl font-semibold leading-none tracking-tight"
          >
            Mot de passe
          </h3>
          <p
            class="text-sm text-muted-foreground"
          >
            Lorem ipsum dolor sit amet, consectetur adipisicing elit. Molestiae, natus!
          </p>
        </div>
        <form>
          <div
            class="p-6 pt-0 flex flex-col gap-4"
          >
            <div
              class="space-y-2"
            >
              <label
                class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
                for="currentPassword"
              >
                Ancien mot de passe
              </label>
              <input
                aria-describedby=":r0:-form-item-description"
                aria-invalid="false"
                class="flex h-10 rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 mt-3 w-80"
                id=":r0:-form-item"
                name="oldPassword"
                placeholder="•••••••••••••"
                type="password"
                value=""
              />
            </div>
            <div
              class="space-y-2"
            >
              <label
                class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
                for="password"
              >
                Nouveau mot de passe
              </label>
              <input
                aria-describedby=":r1:-form-item-description"
                aria-invalid="false"
                class="flex h-10 rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 mt-3 w-80"
                id=":r1:-form-item"
                name="newPassword"
                placeholder="•••••••••••••"
                type="password"
                value=""
              />
            </div>
            <div
              class="space-y-2"
            >
              <label
                class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
                for="confirmPassword"
              >
                Confirmation du mot de passe
              </label>
              <input
                aria-describedby=":r2:-form-item-description"
                aria-invalid="false"
                class="flex h-10 rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 mt-3 w-80"
                id=":r2:-form-item"
                name="confirmNewPassword"
                placeholder="•••••••••••••"
                type="password"
                value=""
              />
            </div>
          </div>
          <div
            class="flex items-center p-6 border-t border-border px-6 py-2"
          >
            <div
              class="flex justify-between items-center w-full"
            >
              <p
                class="text-sm text-muted-foreground"
              >
                Le mot de passe doit au moins contenir 8 caractères.
              </p>
              <button
                class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-semibold ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
                type="submit"
              >
                Sauvegarder
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="rounded-lg border border-border bg-card text-card-foreground shadow-sm"
    >
      <div
        class="flex flex-col space-y-1.5 p-6"
      >
        <h3
          class="text-2xl font-semibold leading-none tracking-tight"
        >
          Mot de passe
        </h3>
        <p
          class="text-sm text-muted-foreground"
        >
          Lorem ipsum dolor sit amet, consectetur adipisicing elit. Molestiae, natus!
        </p>
      </div>
      <form>
        <div
          class="p-6 pt-0 flex flex-col gap-4"
        >
          <div
            class="space-y-2"
          >
            <label
              class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
              for="currentPassword"
            >
              Ancien mot de passe
            </label>
            <input
              aria-describedby=":r0:-form-item-description"
              aria-invalid="false"
              class="flex h-10 rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 mt-3 w-80"
              id=":r0:-form-item"
              name="oldPassword"
              placeholder="•••••••••••••"
              type="password"
              value=""
            />
          </div>
          <div
            class="space-y-2"
          >
            <label
              class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
              for="password"
            >
              Nouveau mot de passe
            </label>
            <input
              aria-describedby=":r1:-form-item-description"
              aria-invalid="false"
              class="flex h-10 rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 mt-3 w-80"
              id=":r1:-form-item"
              name="newPassword"
              placeholder="•••••••••••••"
              type="password"
              value=""
            />
          </div>
          <div
            class="space-y-2"
          >
            <label
              class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
              for="confirmPassword"
            >
              Confirmation du mot de passe
            </label>
            <input
              aria-describedby=":r2:-form-item-description"
              aria-invalid="false"
              class="flex h-10 rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 mt-3 w-80"
              id=":r2:-form-item"
              name="confirmNewPassword"
              placeholder="•••••••••••••"
              type="password"
              value=""
            />
          </div>
        </div>
        <div
          class="flex items-center p-6 border-t border-border px-6 py-2"
        >
          <div
            class="flex justify-between items-center w-full"
          >
            <p
              class="text-sm text-muted-foreground"
            >
              Le mot de passe doit au moins contenir 8 caractères.
            </p>
            <button
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-semibold ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
              type="submit"
            >
              Sauvegarder
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
